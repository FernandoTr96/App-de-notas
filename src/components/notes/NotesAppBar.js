import React from 'react'
import { useSelector, useDispatch } from 'react-redux'
import { DateTime } from 'luxon'
import { updateNote,uploadImage, deleteNote } from '../../actions/notesActions';
import {Loader} from '../loading/Loader'

export const NotesAppBar = ({noteListener}) => {

    const dispatch = useDispatch();
    const { notes: { active }, loader: {loading} } = useSelector(state => state);


    //fecha 
    const config = { month: 'long', day: 'numeric', year: 'numeric' };
    const date = DateTime.fromMillis(active.date).setLocale('mx').toLocaleString(config);


    const handleSave = () => {
        dispatch(
            updateNote(
                {
                    id:active.id,
                    ...noteListener,
                    date: active.date,
                    public_id: active.public_id,
                    url: active.url
                }
            )
        );
    }

    const handleImage = (e)=>{
        let file = e.target.files[0];
        dispatch(uploadImage(file));
    }

    const handleDelete = ()=>{
        dispatch(deleteNote(active.id));
    }

    return (
        <div className='notes__appbar'>
            {active && <span><i className="fas fa-calendar-day"></i> {date} </span>}
            <div>

                <input 
                id='file'
                type='file' 
                className='none'
                accept='*/img'
                onChange={handleImage}
                />

                {loading && <Loader src='https://samherbert.net/svg-loaders/svg-loaders/three-dots.svg' w = {10} h = {10} />}
                <button className='btn-picture'>
                   <label htmlFor='file'>
                        <i className="far fa-images"></i> picture
                   </label>
                </button>

                <button onClick={handleSave} className='btn-save'>
                    <i className="fas fa-save"></i> save
                </button>

                <button onClick={handleDelete}  className='btn-delete'>
                    <i className="fas fa-trash-alt"></i> delete
                </button>

            </div>
        </div>
    )
}
