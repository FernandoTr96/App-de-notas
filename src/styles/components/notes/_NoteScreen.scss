@import '../../_classes/buttons.scss';
@import '../../_classes/main_content.scss';
@import '../../_mixins/mq_breakpoints.scss';

.notes__main-content{

    @extend %main-content;

    .notes__appbar{
        align-items: center;
        background-color: $primary;
        color: $secondary;
        display: flex;
        justify-content: space-between;
        padding: .8rem 3rem;
    
        span{
            font-size: .7rem;
            font-weight: bold;
            opacity: .7;
        }
    
        .btn-picture,.btn-save{
            @extend %buttonIcon;
            padding: .3rem;
            margin-right: .5rem;
            border-radius: .5em;
    
            &:hover{
                transition: background 200ms;
                background-color: rgba(0, 0, 0, 0.158);
            }

            & > *{
                cursor: pointer;
            }
        }

        .btn-delete{
            @extend %buttonIcon;
            padding: .3rem;
            margin-right: .5rem;
            border-radius: .5em;
            &:hover{
                transition: background 200ms;
                background-color: rgb(187, 70, 70)
            }
        }
    
    }

    .notes__content{
        
        @extend %main-content;

        input,textarea{
            border: none;
            outline: 0;
            padding:  0rem 3rem;
            background-color: transparent;
            opacity: 1;
        }

        input{
            color: $primary;
            font-weight: bold;
            text-transform: uppercase;
            font-family: sans-serif;
            text-align: center;
            flex: .2 .2 auto;
        }

        textarea{
            resize: none;
            flex: .7 .7 auto;
            font-family: cursive;
            text-align: justify;
        }

        .notes__image{
            width: 100%;
            padding: 1rem;
        }

        .notes__image img{
            width: 45%;
            height: auto;
            object-fit: cover;
            display: block;
            margin: auto;
            border-radius: .5em;

            @include desktop{
                width: 30%;
            }
        }
    }

}

