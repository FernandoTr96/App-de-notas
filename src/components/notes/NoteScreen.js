import React, { useEffect } from 'react'
import { useSelector } from 'react-redux'
import { NotesAppBar } from './NotesAppBar'
import { useForm } from 'react-hook-form'


export const NoteScreen = () => {

    const { notes: { active } } = useSelector(state => state);

    const { register, watch, setValue } = useForm();

    const noteListener = {
        title: watch('title'),
        body: watch('body')
    }

    //solo cambiar el value cuando cambie la nota activa
    useEffect(() => {
        setValue('title', active.title);
        setValue('body', active.body);
    
    }, [active, setValue])

    
    return (
        <div className='notes__main-content'>
            <NotesAppBar noteListener={noteListener} />
            <div className='notes__content'>

                <input
                    type='text'
                    placeholder='Title'
                    autoComplete='off'
                    {...register('title')}
                    defaultValue={watch('title')}
                />

                <textarea
                    placeholder='What happened today ?'
                    {...register('body')}
                    defaultValue={watch('body')}
                ></textarea>

                {
                    active?.url &&
                    <div className='notes__image'>
                        <img
                            src={active.url}
                            alt='journal'
                        />
                    </div>
                }

            </div>
        </div>
    )
}
