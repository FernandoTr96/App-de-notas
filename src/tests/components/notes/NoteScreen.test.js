import React from 'react';
import {mount} from 'enzyme'
import configureStore from 'redux-mock-store'
import {MemoryRouter} from 'react-router-dom'
import thunk from 'redux-thunk'
import { Provider } from 'react-redux';
import { NoteScreen } from '../../../components/notes/NoteScreen';

const middlewares = [thunk];
const mockstore = configureStore(middlewares);
const init = {
    loader:{
        loading: false
    },
    notes:{
        active:{
            title: 'TITLERANDOM',
            body: 'BODYRANDOM',
            date: 1634420927846
        }
    }
};
const store = mockstore(init);


const wrapper = mount(
    <Provider store = {store}>
        <MemoryRouter>
            <NoteScreen/>
        </MemoryRouter>
    </Provider>
);

describe('Pruebas en <NoteScreen/>', () => {
    
    test('debe pintarse correctamente', () => {
        expect(wrapper).toMatchSnapshot(); 
    });

    test('debe cambiar el valor de los input si la nota activa cambia', () => {
        
        expect(wrapper.find('input[type="text"]').prop('defaultValue')).toBe(init.notes.active.title);
        expect(wrapper.find('textarea').prop('defaultValue')).toBe(init.notes.active.body);
    
    });

    test('debe mostrar una imagen si es que tiene', () => {
        
        const storeWithImg = mockstore({
            ...init,
            notes:{
                active:{
                   ...init.notes.active,
                   url: 'https://placeimg.com/100/100/any'
                }
            }
        })

        const wrapper = mount(
            <Provider store={storeWithImg} >
                <MemoryRouter>
                    <NoteScreen/>
                </MemoryRouter>
            </Provider>
        );

        expect(wrapper.find('.notes__image').exists()).toBe(true);

    });

});