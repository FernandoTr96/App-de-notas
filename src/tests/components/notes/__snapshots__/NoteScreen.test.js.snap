// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <NoteScreen/> debe pintarse correctamente 1`] = `
<div
  className="notes__main-content"
>
  <div
    className="notes__appbar"
  >
    <span>
      <i
        className="fas fa-calendar-day"
      />
       
      16 de octubre de 2021
       
    </span>
    <div>
      <input
        accept="*/img"
        className="none"
        id="file"
        onChange={[Function]}
        type="file"
      />
      <button
        className="btn-picture"
      >
        <label
          htmlFor="file"
        >
          <i
            className="far fa-images"
          />
           picture
        </label>
      </button>
      <button
        className="btn-save"
        onClick={[Function]}
      >
        <i
          className="fas fa-save"
        />
         save
      </button>
      <button
        className="btn-delete"
        onClick={[Function]}
      >
        <i
          className="fas fa-trash-alt"
        />
         delete
      </button>
    </div>
  </div>
  <div
    className="notes__content"
  >
    <input
      autoComplete="off"
      defaultValue="TITLERANDOM"
      name="title"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="Title"
      type="text"
    />
    <textarea
      defaultValue="BODYRANDOM"
      name="body"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="What happened today ?"
    />
  </div>
</div>
`;
